### PR TITLE
Introduce Node Pattern tutorial

### PR DESCRIPTION
**Pull request by jonatas**
I'm really excited about work with node pattern and I bring this extra tutorial about how to interpret and create node patterns and learn a bit about this small compiler.

Suggestions are welcome!

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* (n/a) Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* (n/a) Added tests.
* (n/a) Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* (n/a) The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* (n/a) Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by jonatas** [comment]
I'm not sure how to visualize it in my computer. I just followed the markdown but I'm very curious about run and preview the manual in my local machine.

**Comment by bbatsov** [comment]
All you need is http://www.mkdocs.org/ locally.

**Comment by bbatsov** [comment]
https://rubocop.readthedocs.io/en/latest/contributing/#working-on-the-manual

File: manual/node_pattern.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+```
+$ ruby-parse -e '1'
+(int 1)
+```
+
+- `int` will match exactly the node
+- `(int 1)` will match precisely the node
+
+## `_` for any value
```
Comment (pirj) [inline manual/node_pattern.md]: "any single node" probably?

File: manual/node_pattern.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+node.int_type? && node.children.size == 1
+```
+
+while `(int _ _)` will be translated to something like:
+
+```ruby
+node.int_type? && node.children.size == 2
+```
+
+## `...` for something else
```
Comment (pirj) [inline manual/node_pattern.md]: "several subsequent nodes"?

File: manual/node_pattern.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+While `_` limits the size, `...` make it open for more children.
```
Comment (pirj) [inline manual/node_pattern.md]: What do you think of the following?
> Where `_` matches (captures?) a single node, `...` eagerly matches one or more subsequent nodes.

File: manual/node_pattern.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+And the example `(int ...)` will be translated into something like:
```
Comment (pirj) [inline manual/node_pattern.md]: `int` is definitely a bad example here, since there's just a single child node. `send` with couple parameters?

File: manual/node_pattern.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+And it matches with any number as well.
+
+It's useful when you want to check some variable internal nodes but with a
+final with the same results. For example, let's use a classic `something.save`.
+
+We can also have `person.save` or `person.address.save`, and we want to match
+both. So, let's check how it looks like in the AST:
+
+```
+$ ruby-parse -e 'person.save'
```
Comment (pirj) [inline manual/node_pattern.md]: `sum(1, 2)` 
```
(send nil :sum
  (int 1)
  (int 2))
```

would be easier to understand than a nested save.

File: manual/node_pattern.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+  (send
+    (send nil :person) :address) :save)
+```
+
+The expression will not match, and it can be `a.b.c.d.save`.
+
+In such cases you can use `...` to match both cases:
+
+```
+(send ... :save)
```
Comment (pirj) [inline manual/node_pattern.md]: I think you can equally use `(send _ :save)` in this case, so it's a bad example for unique use of `...`.

I see primary use of `...` to match subsequent, not nested, elements.

File: manual/node_pattern.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+Lets make it a bit more complex and introduce floats:
+
+```
+$ ruby-parse -e '1'
+(int 1)
+$ ruby-parse -e '1.0'
+(float 1.0)
+```
+
+- `({int float} _)` - int or float types, no matter the value
```
Comment (pirj) [inline manual/node_pattern.md]: This may be simplified to just `{int float}` I suppose.

I see {a b} as `a || b`.

File: manual/node_pattern.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+## `$` for captures
+
+You can capture elements or nodes along with your search, prefixing the expression
+with `$`. For example, in a tuple like `(int 1)` with a search like `(int $_)`,
+you can catch element if the search matches.
```
Comment (pirj) [inline manual/node_pattern.md]: You provide `(int 1)` as an example, so I don't think `if` applies.

File: manual/node_pattern.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+You can also capture multiple things like:
+
+```
+(${int float} $_)
+```
+
+It will bring an array with `[:int, 1]` for a tuple like (int 1).
```
Comment (pirj) [inline manual/node_pattern.md]: `s/bring/return/`?

File: manual/node_pattern.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+The tuple can be captured using the `$` before the open parens:
+
+```
+$({int float} _)
```
Comment (pirj) [inline manual/node_pattern.md]: Same as `${int float}` I suppose.

File: manual/node_pattern.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+```
+
+## predicate methods
```
Comment (pirj) [inline manual/node_pattern.md]: Capitalization

File: manual/node_pattern.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+You can also use methods that end with question mark `?`, to ask for some
+method in the node. Example:
```
Comment (pirj) [inline manual/node_pattern.md]: RDoc is more detailed here:
>
                    # words which end with a ? are predicate methods,
                    # are are called on the target to see if it matches
                    # any Ruby method which the matched object supports
                    # can be used

File: manual/node_pattern.md [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+- `int_type?` can be used herein replacement of `(int _)`
```
Comment (pirj) [inline manual/node_pattern.md]: It shouldn't. What it can is a side effect of how it's implemented internally, e.g. I assume that Integer AST node class responds to `int_type?`, and as the result, `int` matches, not the other way around.

Comment (pirj) [reply manual/node_pattern.md]: Thing is, `int_type?` is a method of AST, while `odd?` is a method called on that object.

**Comment by mikegee** [comment]
Thanks for working on this. I don't know how to use this stuff. 😄

File: manual/development.md [hunk 14]
```diff
@@ -95,6 +95,8 @@ Now that you're familiar with AST, you can learn a bit about the
 [node pattern](http://www.rubydoc.info/gems/rubocop/RuboCop/NodePattern)
 and use patterns to match with specific nodes that you want to match.
 
+You can learn more about [Node Pattern here](https://rubocop.readthedocs.io/en/latest/node_pattern/).
```
Comment (bbatsov) [inline manual/development.md]: Only `here` should be the target of this link.

File: manual/node_pattern.md [hunk 15]
```diff
@@ -0,0 +1,270 @@
+# Node Pattern
+
+Node pattern is a small language to help find specific nodes in
```
Comment (bbatsov) [inline manual/node_pattern.md]: small language -> DSL :-)

File: manual/node_pattern.md [hunk 16]
```diff
@@ -0,0 +1,270 @@
+# Node Pattern
```
Comment (bbatsov) [inline manual/node_pattern.md]: I think this should tutorial should begin with a description of the problem node patterns were created to solve - a few examples of "legacy code" that had to do multiple node destructurings to access some node(s) and the same code expressed using node patterns. You can find plenty of examples of this in RuboCop's own codebase - after the node pattern was introduced by @alexdowad we've started gradually porting cops to it. Just look at the git history.

Comment (jonatas) [reply manual/node_pattern.md]: Good point. I'll get a few examples and use it!
I directed to the details of the language, but it's an excellent addition to understanding why it was built.

I'll certainly take a look at the codebase. Do you remember any extraordinary example we can show the "before" and "after" using the Node Pattern?

File: manual/node_pattern.md (hunk 13 above)
Comment (jonatas) [reply manual/node_pattern.md]: Yes. Exactly. I think we need to emphasize it in somehow.

**Comment by jonatas** [comment]
Thanks for the detailed review @pirj. I addressed all the comments (I think).

**Comment by jonatas** [comment]
I'm still thinking about migrating the examples to use `rake repl` and use Ruby to demonstrate all examples. As I did in the Development tutorial. Not sure how valuable it can be. The idea of just open the terminal, start copying and pasting from the examples are generally comfortable and engage more the reader. At least me as a reader.

Do you agree about code examples?

**Comment by alexdowad** [comment]
I'm surprised and honored that someone liked my code enough to write this much about it. Thanks for the tutorial, @jonatas!

I don't think it would hurt to add a note (at the end) that the source code is easy to read and hack on, if that is true. Anyways, it's your tutorial, so please don't feel you need to take my suggestions.

**Comment by jonatas** [comment]
Thanks, @alexdowad! I really enjoyed the Node Pattern approach. I started my own [toy](https://github.com/jonatas/fast) independent of RuboCop and it's an interesting learning experience for me. Your code was my major reference there 👍 

Added the message to encourage more people to hack it ✅

File: manual/node_pattern.md [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+$ ruby-parse -e '1.0'
+(float 1.0)
+```
+
+- `({int float} _)` - int or float types, no matter the value
+
+## `$` for captures
+
+You can capture elements or nodes along with your search, prefixing the expression
+with `$`. For example, in a tuple like `(int 1)`, you can capture the value using `(int $_)`.
```
Comment (pirj) [inline manual/node_pattern.md]: Mind providing a code example?
```
> node = RuboCop::ProcessedSource.new('1', 2.3).ast
> puts node 
=> (int 1)
> pattern = RuboCop::NodePattern.new('(int $_)')
> pattern.match node # => 1
```

File: manual/node_pattern.md [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+
+You can also capture multiple things like:
+
+```
+(${int float} $_)
+```
+
+It will return an array with `[:int, 1]` for a tuple like (int 1).
```
Comment (pirj) [inline manual/node_pattern.md]: Mind elaborating what is `it` in this context?

File: manual/node_pattern.md [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+The tuple can be entirely captured using the `$` before the open parens:
+
+```
+$({int float} _)
+```
+
+Or remove the parens and match directly from node head:
+
+```
+${int float}
```
Comment (pirj) [inline manual/node_pattern.md]: This is a consequence that `(int _)` works in the same way as `int`, and `$({_ _})` as `${(...) (...)}`, but it's hard to comprehend it just at this point of tutorial.

File: manual/node_pattern.md [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+```
+
+## Predicate methods
+
+Words which end with a `?` are predicate methods, are called on the target
+to see if it matches any Ruby method which the matched object supports can be
+used.
```
Comment (pirj) [inline manual/node_pattern.md]: I know it's coming directly from code doc, but it's very hard to follow.

File: manual/node_pattern.md [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+
+Example:
+
+- `int_type?` can be used herein replacement of `(int _)`.
+
+And refactoring the expression to allow both int or float types:
+
+- `{int_type? float_type?}` can be used herein replacement of `({int float} _)`
```
Comment (pirj) [inline manual/node_pattern.md]: Why would we refactor `{int float}` to `{int_type? float_type?}`?

Let's probably split the topic related to generic predicate methods (e.g.`empty?`, `nil?`, `zero?`) from topic about inferred types and send-type?

File: manual/node_pattern.md [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+
+You can also use it at the node level, asking for each child:
+
+- `(int odd?)` will match only with odd numbers, asking it to the current
+  number.
```
Comment (pirj) [inline manual/node_pattern.md]: Why is it referred to as "current"? Probably "matched"?

File: manual/node_pattern.md [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+
+## `[]` for or `<expression>`
```
Comment (pirj) [inline manual/node_pattern.md]: "and", conjunction

File: manual/node_pattern.md [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+
+Imagine you want to check if the number is `odd?` and also not `zero?`:
```
Comment (pirj) [inline manual/node_pattern.md]: `positive?` would fit better here, as it does not need negation to be used together with `odd?` to bring some results.

File: manual/node_pattern.md [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+
+`(int [odd? !zero?])` - is an int and the value should be odd and not zero.
```
Comment (pirj) [inline manual/node_pattern.md]: e.g. `(int [odd? positive?])` will match `1`, while won't match`-1`, `-2`, `2`.

File: manual/node_pattern.md [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+
+Sometimes, we want to add an extra logic. Let's imagine we're searching for
```
Comment (pirj) [inline manual/node_pattern.md]: s/an//

File: manual/node_pattern.md [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+prime numbers, so we have a method to define it:
```
Comment (pirj) [inline manual/node_pattern.md]: Won't "and we have a method to detect it" will be more descriptive?

File: manual/node_pattern.md [hunk 28]
```diff
@@ -0,0 +1,292 @@
...
+
+```ruby
+def prime?(n)
+  if n <= 1
+    false
+  elsif n == 2
+    true
+  else
+    (2..n/2).none? { |i| n % i == 0}
```
Comment (pirj) [inline manual/node_pattern.md]: Space before closing curly brace missing.

File: manual/node_pattern.md [hunk 29]
```diff
@@ -0,0 +1,292 @@
...
+  end
+end
+```
+
+We can incorporate this using `#prime?` in the expression:
```
Comment (pirj) [inline manual/node_pattern.md]: How about "we can use the `#prime?` method directly in the expression"?

File: manual/node_pattern.md [hunk 30]
```diff
@@ -0,0 +1,292 @@
...
+
+```
+(int [odd? !zero? #prime?])
+```
+
+It will match only `odd?`, non zero and prime numbers.
```
Comment (pirj) [inline manual/node_pattern.md]: Primes are not zero ( and also positive), and the only even one is `2`.
You might say that it will match all prime numbers except `2`.

File: manual/node_pattern.md [hunk 31]
```diff
@@ -0,0 +1,292 @@
...
+- `(int _)` will match any number
+- `(int _ _)` will not match because `int` types have just one child that
+  contains the value.
+
+
+## `...` for several subsequent nodes
+
+Where `_` matches a single node, `...` eagerly matches one or more subsequent nodes.
+
+It's useful when you want to check some variable internal nodes but with a
```
[the remark this answers is not in the document]
Comment (jonatas) [reply manual/node_pattern.md]: Great catch! fixed.

File: manual/node_pattern.md (hunk 17 above)
Comment (jonatas) [reply manual/node_pattern.md]: I'm thinking about improving all of them using code.

Thanks for the sample: 🤗

File: manual/node_pattern.md (hunk 18 above)
Comment (jonatas) [reply manual/node_pattern.md]: Sure thing. Fixed.

File: manual/node_pattern.md (hunk 30 above)
Comment (jonatas) [reply manual/node_pattern.md]: I'll remove the extra logic and just keep the `prime` part.

File: manual/node_pattern.md (hunk 27 above)
Comment (jonatas) [reply manual/node_pattern.md]: Sure thing. Fixed.

**Comment by bbatsov** [comment]
There's certainly more room for improvement, but I'll merge this at this point as I plan to cut a new RuboCop release later today. 

Thanks for your contribution, @jonatas!

**Comment by jonatas** [comment]
Great @bbatsov! Glad to contribute :)
